skbio/sequence/_interval.py

```diff
@@ -0,0 +1,127 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+from __future__ import absolute_import, division, print_function
+
+from ._feature import Feature
+from .intersection import Interval, IntervalTree
+import itertools
+from collections import defaultdict
+
+class IntervalMetadata():
+    ''' Store metadata corresponding to Features and Intervals.
+
+    There are two underlying data structures namely
+    1. An IntervalTree (interval_tree)
+    2. A bidirectional hash table (feature_map)
+
+    Attributes
+    ----------
+    interval_tree
+    feature_table
+    interval_table
+    interval_ids
+
+    '''
+    def __init__(self):
+        self.interval_tree = IntervalTree()
+        # maps features to intervals and vice versa
+        self.feature_table = defaultdict(list)
+
+    def concat(self, x, kw='location'):
+        """ Adds a feature to the metadata object.
+
+        Parameters
+        ----------
+        x : skbio.sequence.feature
+            The feature object being added.
+        kw : str
+            The keyword associated with the intervals.
+
+        Note
+        ----
+        The intervals associated with a feature is assumed to be under
+        the `location` keyword by default.
+        """
+        if not isinstance(x, Feature):
+            raise ValueError('x is not an instance of `Feature`')
+
+        intervals = x[kw]
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for interval in intervals:
+            self.interval_tree.add(interval[0], interval[1], x)
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for k, v in x.items():
+            if x not in self.feature_table:
+                self.feature_table[(k, v)] = [x]
+            else:
+                self.feature_table[(k, v)].append(x)
+
+
+    def query(self, **kwargs):
+        """
```

Comment:
:+1:  That's a nice way to integrate accepting intervals.

On Fri, Feb 5, 2016 at 12:41 PM, Zech Xu <notifications@github.com> wrote:

> In skbio/sequence/_interval.py
> <https://github.com/RNAer/scikit-bio/pull/1#discussion_r52069884>:
>
> > +        # since this will be using a little more memory, since the whole
> > +        # feature object will be stored.
> > +        for interval in intervals:
> > +            self.interval_tree.add(interval[0], interval[1], x)
> > +        # TODO: The below will require careful consideration
> > +        # since this will be using a little more memory, since the whole
> > +        # feature object will be stored.
> > +        for k, v in x.items():
> > +            if x not in self.feature_table:
> > +                self.feature_table[(k, v)] = [x]
> > +            else:
> > +                self.feature_table[(k, v)].append(x)
> > +
> > +
> > +    def query(self, **kwargs):
> > +        """
>
> I suggest the func API is query(self, *args, **kwargs), so it will
> provide a powerful query by combinations of locations and attributes. args
> accepts a list of location intervals and kwargs accepts keyword arg of
> attributes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RNAer/scikit-bio/pull/1/files#r52069884>.
>